sw/host/opentitanlib/src/io/gpio.rs

```diff
@@ -200,3 +209,36 @@ pub trait GpioMonitoring {
         continue_monitoring: bool,
     ) -> Result<MonitoringReadResponse>;
 }
+
+/// Represents one entry in the specification of a bitbanging waveform.
+pub enum BitbangEntry<'rd, 'wr> {
+    /// Represents a sequence of pin values.  Bit 0 in each byte controls the first `GpioPin` in
+    /// the slice given to `GpioBitbanging::run()`, bit 1 controls the second `GpioPin`, and so
+    /// on.  Each byte is applied to the pins in sequence, with a particular delay between each
+    /// given by the `clock_tick` argument to `run()`.
+    Write(&'wr [u8]),
+    /// Represents a sequence of pin values as above, but additionally captures the value of any
+    /// pins in `Input` or `OpenDrain` mode.  At each clock tick, the input levels are sampled
+    /// just before the given output levels are applied, meaning that the data coming back will be
+    /// offset by one sample.
+    Both(&'wr [u8], &'rd mut [u8]),
+    /// Represents a delay of the given number of clock ticks in which the output levels are held
+    /// as indicated by the last byte of the preceeding `Write`/`Both` entry.
+    ///
+    /// A delay shorter than one tick is invalid.  A delay of one tick is equivalent to not
```

Comment:
Can you document the unit of the delay, and perhaps what is the range accepted?

Reply:
It is documented above "a delay of the given number of clock ticks", so the range is 1 clock tick to 2^32-1 clock ticks, (which would be around ten hours assuming a maximum frequency of 115.2kHz).